### ISSUE TITLE
jax-rs json mapping date to XMLGregorianCalendar incorrectly

### ISSUE THREAD

**Issue by nadiramra** [OpenLiberty/open-liberty#3504]
Hi, 

My server is set to CDT timezone, I do not understand when a JSON request comes in to  JAX-RS web service with date field such as:

{
  "dateField": "2017-05-08"
}

And the dateField is defined to be a type of XMLGregorianCalendar, Liberty seems to set XMLGregorianCalendar field to 2017-05-07T19:00:00.000-05:00.   The problem I have with Liberty/Jackson is that it is treating the date as a timestamp and changing the date, when it should not. It is a date.  The XMLGregorianCalendar allows you to indicate that time fields are not specified.   

If I send the same date in XML format:

<dateField>2017-05-08</dateField>

I get what I expect, the value:  2017-05-08.

This is the same REST web service that can return either XML or JSON.   It is very frustrating.   And yes, I know I can put the burden on myself and package Jackson and override deserializer, etc, but why should I have to do that?!  In my mind there should be a consistency.   Liberty should be doing the right thing, and in this case treating the date as a timestamp is not right.

**Comment by andymc12** [comment]
Hi @nadiramra ,

There are a few options to try:
1) Package Jackson (or some other object-to-JSON serialization library) and configure it - as you mentioned, it doesn't sound like you want to try this approach.  Still, others who read this may prefer this approach, so I mention it here.  Instructions for packaging Jackson with your JAX-RS application are [here](https://developer.ibm.com/wasdev/docs/jax-rs-websphere-liberty-byo-jackson/).

2) You could try this using JAX-RS 2.1 (currently in beta, but delivering GA soon).  JAX-RS 2.1 (per the spec) uses JSON-B to serialize/deserialize to/from JSON.  I don't recall the specifics offhand, but JSON-B does differ from Jackson.  Alex Theedom wrote a great series on JSON Binding, and in [part 4](https://www.ibm.com/developerworks/library/j-javaee8-json-binding-4/index.html) he specifically compares Jackson with JSON-B, including processing of Date types.

2.1) If the JSON-B that comes out of the box with JAX-RS 2.1 doesn't work quite the way you want it to, then you can add the `jsonb-1.0` feature and then you will have access to all of the `@JsonbDateFormat`, etc. APIs for customizing you JSON schema.

3) Lastly, you could create a pull request that configures OpenLiberty's internal version of Jackson.  The place where we register and configure the Jackson provider for JAX-RS 2.0 is [here](https://github.com/OpenLiberty/open-liberty/blob/7c9c9f371239de834ca338e45ff66cba3a495209/dev/com.ibm.ws.jaxrs.2.0.common/src/org/apache/cxf/jaxrs/provider/ProviderFactory.java#L309).

If you have any questions, please let me know.  Thanks, Andy

**Comment by nadiramra** [comment]
Hi Andy, thanks for the response.  I know there are various options one could use (thanks for listing some I did not know about!).  The main point is that this is a bug, no?  And this should be fixed.  I do not understand why we do not fix obvious defects.

**Comment by nadiramra** [comment]
Andy, would do you mean "...create a pull request..."?

**Comment by andymc12** [comment]
Hi @nadiramra - it does seem like a bug to me, but I have not investigated it sufficiently myself.  If there is an obvious fix for this, then I would recommend that you propose it (pull request means that you have a patch that you request gets merged into the code base - more info on contributing to OpenLiberty [here](https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Development)) with a patch (pull request).

The open source community relies on contributions from the community - and in OpenLiberty, we definitely want to encourage community contributions. 

You could probably implement this by using the `SerializationConfig` setting, `WRITE_DATES_AS_TIMESTAMPS `.  The Jackson source code for `SerializationConfig` is [here](https://github.com/codehaus/jackson/blob/master/src/mapper/java/org/codehaus/jackson/map/SerializationConfig.java).  So maybe something like:
```
DateFormat df = AccessController.doPrivileged(new PrivilegedAction<DateFormat>() {

            @Override
            public DateFormat run() {
                String format = System.getProperty("io.openliberty.jaxrs.jackson.dateFormat");
                if (format == null) {
                     return null;
                }
                return new SimpleDateFormat(format);
            }
        });
serializationConfig = serializationConfig.withDateFormat(df);
```

Hope this helps, Andy

**Comment by andymc12** [comment]
Per our discussion on July 30, 2018, Nadir decided that he would continue to use JAX-RS 1.1.  We agreed that the serialization/deserialization process is inconsistent between XML (using JAXB) and JSON (using Jackson), but that the process should be controlled by the application code.  In JAX-RS 2.1 (now available as of Liberty 18.0.0.2), users can control the serialization/deserialization between JSON/Objects via JSON-B which is built-in to the product.  This should be the future direction for this issue.
